### PR TITLE
Core: Update TableMetadataParser to ensure all streams closed

### PR DESCRIPTION
**Pull request by erik-grepr**
Minor try-with-resources tweak to fix an unclosed stream memory leak I encountered.

More specifically, this occurs when the file is Gzipped and file.newStream().read() throws an exception.

The `GZIPInputStream` constructor calls read on it's input to load the header.

see: https://stackoverflow.com/questions/12552863/correct-idiom-for-managing-multiple-chained-resources-in-try-with-resources-bloc

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/TableMetadataParser.java [hunk 1]
```diff
@@ -122,13 +122,23 @@ public static void write(TableMetadata metadata, OutputFile outputFile) {
...
+      if (isGzip) {
+        try (OutputStream gos = new GZIPOutputStream(os);
+            OutputStreamWriter osw = new OutputStreamWriter(gos, StandardCharsets.UTF_8)) {
+          JsonGenerator generator = JsonUtil.factory().createGenerator(os);
+          generator.useDefaultPrettyPrinter();
+          toJson(metadata, generator);
+          generator.flush();
+        }
+      } else {
+        try (OutputStreamWriter osw = new OutputStreamWriter(os, StandardCharsets.UTF_8)) {
```
Comment (findepi) [inline core/src/main/java/org/apache/iceberg/TableMetadataParser.java:135]: I am not convinced we need the two code branches for this, nor that it's desired.
`OutputStream::close` must obey general `Closeable::close` contract which requries the implementation to be idempotent.

Thus, if we had

```
OutputStream stream = overwrite ? outputFile.createOrOverwrite() : outputFile.create();
    try (OutputStream ou = isGzip ? new GZIPOutputStream(stream) : stream;
```

we now should be able to:

```
try (OutputStream stream = overwrite ? outputFile.createOrOverwrite() : outputFile.create();
   OutputStream ou = isGzip ? new GZIPOutputStream(stream) : stream;
```

guaranteeing all streams to be closed, regardless at which point an exception is thrown.

Can you try that?

**Review by findepi — APPROVED** [review]

File: core/src/main/java/org/apache/iceberg/TableMetadataParser.java (hunk 1 above)
Comment (findepi) [reply core/src/main/java/org/apache/iceberg/TableMetadataParser.java]: thanks!

> Is there a gradle target to auto-format?

```
./gradlew -DallModules=true spotlessApply
```

( see also https://gist.github.com/findepi/da2ea85d095eaa0f3cfdcbb311ff59ed#reformat-code )

**Comment by findepi** [comment]
@erik-grepr can you please review the build results?

**Comment by erik-grepr** [comment]
Should be good to re-run the workflows.

**Review by findepi — APPROVED** [review]

**Review by dramaticlly — APPROVED** [review]

**Comment by findepi** [comment]
Merged, thanks!

**Pull request merged**